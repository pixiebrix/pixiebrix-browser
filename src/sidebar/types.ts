/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type RegistryId, type UUID } from "@/core";
import { type FormDefinition } from "@/blocks/transformers/ephemeralForm/formTypes";
import { type RendererPayload } from "@/runtime/runtimeTypes";

export type RendererError = {
  /**
   * A unique id for the content, used control re-rendering (similar to `key` in React)
   */
  key: string;
  /**
   * The error message to show in the panel
   */
  // TypeScript was having problems handling the type SerializedError here
  error: unknown;
  /**
   * The extension run id.
   * @since 1.7.0
   */
  runId: UUID;
  /**
   * The extension id that produced the error
   * @since 1.7.0
   */
  extensionId: UUID;
};

/**
 * Entry types supported by the sidebar.
 *
 * Current supports panels and ephemeral forms. In the future we may also support button entries, etc.
 *
 * @see PanelEntry
 * @see FormEntry
 */
export type EntryType = "panel" | "form" | "temporaryPanel";

/**
 * The information required to run the renderer of a pipeline, or error information if the pipeline run errored.
 */
export type PanelPayload = RendererPayload | RendererError | null;

type BasePanelEntry = {
  /**
   * The id of the extension that added the panel
   */
  extensionId: UUID;
  /**
   * Heading for tab name in the sidebar
   */
  heading: string;
  /**
   * The information required to run the renderer of a pipeline, or error information if the pipeline run errored.
   */
  payload: PanelPayload;
};

/**
 * A panel added by an extension attached to an SidebarExtensionPoint
 * @see SidebarExtensionPoint
 */
export type PanelEntry = BasePanelEntry & {
  /**
   * The blueprint associated with the extension that added the panel.
   *
   * Used to give preference to blueprint side panels when using the "Show Sidebar" brick.
   *
   * @since 1.6.5
   */
  blueprintId: RegistryId | null;
  /**
   * The sidebar extension point
   * @see SidebarExtensionPoint
   */
  extensionPointId: RegistryId;
};

/**
 * An ephemeral panel to show in the sidebar. Only one temporary panel can be shown from an extension at a time.
 */
export type TemporaryPanelEntry = BasePanelEntry & {
  /**
   * Unique identifier for the temporary panel instance. Used to correlate panel-close action.
   */
  nonce: UUID;
};

/**
 * An ephemeral form to show in the sidebar. Only one form can be shown from an extension at a time.
 * @see ModalTransformer
 */
export type FormEntry = {
  /**
   * The extension that created the form
   */
  extensionId: UUID;
  /**
   * Unique identifier for the form instance. Used to correlate form submission/cancellation.
   */
  nonce: UUID;
  /**
   * The form schema and configuration
   */
  form: FormDefinition;
};

/**
 * The entries currently added to the sidebar
 */
export type SidebarEntries = {
  panels: PanelEntry[];
  forms: FormEntry[];
  temporaryPanels: TemporaryPanelEntry[];
};

/**
 * A request to activate a panel in the sidebar
 * @since 1.6.5
 */
export type ActivatePanelOptions = {
  /**
   * Force-activate the panel, even if the user is currently viewing a different panel that doesn't match the criteria
   *
   * @since 1.6.5
   */
  force?: boolean;

  /**
   * Refresh the panel content (default=true).
   *
   * Has no effect if the sidebar is not already showing
   *
   * @since 1.7.0
   */
  refresh?: boolean;

  /**
   * The id of the extension panel to show. Included so the Page Editor can request a specific panel to show when
   * editing the extension
   *
   * @since 1.6.5
   */
  extensionId?: UUID;
  /**
   * The blueprint of the extension panel to show
   *
   * @since 1.6.5
   */
  blueprintId?: RegistryId;
  /**
   * A panel heading name to match
   *
   * @since 1.6.5
   */
  panelHeading?: string;
};

/**
 * Metadata about the extension that produced the panel content
 * @since 1.7.0
 */
export type PanelRunMeta = {
  runId: UUID;
  extensionId: UUID;
};
