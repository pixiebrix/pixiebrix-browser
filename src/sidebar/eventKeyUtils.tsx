/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import type {
  SidebarEntries,
  SidebarEntry,
  SidebarState,
} from "@/types/sidebarTypes";
import {
  isModActivationPanelEntry,
  isPanelEntry,
  isStaticPanelEntry,
} from "@/types/sidebarTypes";
import hash from "object-hash";
import { sortBy } from "lodash";
import type { Nullishable } from "@/utils/nullishUtils";

export function eventKeyForEntry(
  entry: Nullishable<SidebarEntry>,
): Nullishable<string> {
  if (entry == null) {
    return null;
  }

  if (isModActivationPanelEntry(entry)) {
    return `activate-${hash(sortBy(entry.mods.map((x) => x.modId)))}`;
  }

  if (isPanelEntry(entry)) {
    return `panel-${entry.extensionId}`;
  }

  if (isStaticPanelEntry(entry)) {
    return `static-${entry.key}-panel`;
  }

  // Use nonce to keep eventKeys unique for forms and temporary panels from the same extension
  return `${entry.type}-${entry.nonce}`;
}

/**
 * Return the default tab to show.
 *
 * Give preference to:
 * 1. Most recent ephemeral form
 * 2. Most recent temporary panel
 * 3. First panel
 * 4. Mod activation panel
 * 5. Static panels (i.e., the Mod Launcher)
 */
export function defaultEventKey(
  {
    forms = [],
    panels = [],
    temporaryPanels = [],
    staticPanels = [],
    modActivationPanel = null,
  }: SidebarEntries,
  closedTabs: SidebarState["closedTabs"],
): Nullishable<string> {
  if (forms.length > 0) {
    return eventKeyForEntry(forms.at(-1));
  }

  if (temporaryPanels.length > 0) {
    return eventKeyForEntry(temporaryPanels.at(-1));
  }

  const openPanels = getOpenPanelEntries(panels, closedTabs);
  if (openPanels.length > 0) {
    return eventKeyForEntry(openPanels.at(0));
  }

  if (modActivationPanel) {
    return eventKeyForEntry(modActivationPanel);
  }

  const openStaticPanels = getOpenPanelEntries(staticPanels, closedTabs);
  if (openStaticPanels.length > 0) {
    return eventKeyForEntry(openStaticPanels.at(0));
  }

  return null;
}

export function getOpenPanelEntries(
  entries: SidebarEntry[],
  closedTabs: SidebarState["closedTabs"],
): SidebarEntry[] {
  return entries.filter((entry) => {
    const key = eventKeyForEntry(entry);
    // eslint-disable-next-line security/detect-object-injection -- generated by eventKeyForEntry
    return !key || !closedTabs[key];
  });
}
