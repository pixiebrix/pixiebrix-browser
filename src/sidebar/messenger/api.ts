/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* Do not use `registerMethod` in this file */
import { getMethod, getNotifier } from "webext-messenger";

const target = { tabId: "this", page: "/sidebar.html" } as const;
const sidebarInThisTab = {
  renderPanels: getMethod("SIDEBAR_RENDER_PANELS", target),
  activatePanel: getMethod("SIDEBAR_ACTIVATE_PANEL", target),
  showForm: getMethod("SIDEBAR_SHOW_FORM", target),
  hideForm: getMethod("SIDEBAR_HIDE_FORM", target),
  pingSidebar: getMethod("SIDEBAR_PING", target),
  showTemporaryPanel: getMethod("SIDEBAR_SHOW_TEMPORARY_PANEL", target),
  updateTemporaryPanel: getNotifier("SIDEBAR_UPDATE_TEMPORARY_PANEL", target),
  hideTemporaryPanel: getMethod("SIDEBAR_HIDE_TEMPORARY_PANEL", target),
  showModActivationPanel: getMethod("SIDEBAR_SHOW_ACTIVATE_RECIPE", target),
  hideModActivationPanel: getMethod("SIDEBAR_HIDE_ACTIVATE_RECIPE", target),
};

export default sidebarInThisTab;
