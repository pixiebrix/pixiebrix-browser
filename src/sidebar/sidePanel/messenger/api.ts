/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * @file This file defines the public API for the sidePanel, with some
 * exceptions that use `expectContext`. It uses the `messenger/api.ts` name
 * to match that expectation and avoid lint issues.
 */

import { getExtensionConsoleUrl } from "@/utils/extensionUtils";
import {
  DISPLAY_REASON_EXTENSION_CONSOLE,
  DISPLAY_REASON_RESTRICTED_URL,
} from "@/tinyPages/restrictedUrlPopupConstants";
import { isScriptableUrl } from "webext-content-scripts";
import { isObject } from "@/utils/objectUtils";
import { expectContext } from "@/utils/expectContext";
import { type Target } from "webext-messenger";

export function getAssociatedTabId(): number {
  expectContext("sidebar");
  const tabId = new URLSearchParams(window.location.search).get("tabId");
  return Number(tabId);
}

export function getAssociatedTarget(): Target {
  return { tabId: getAssociatedTabId(), frameId: 0 };
}

const PING_MESSAGE = "PING_SIDE_PANEL";
// Do not use the messenger because it doesn't support retry-less messaging
// TODO: Drop after https://github.com/pixiebrix/webext-messenger/issues/59
export function respondToPings() {
  expectContext("sidebar");
  chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
    if (
      isObject(message) &&
      message.type === PING_MESSAGE &&
      sender.tab?.id === getAssociatedTabId()
    ) {
      sendResponse(true);
    }
  });
}

export async function isSidePanelOpen(): Promise<boolean> {
  // Sync check where possible
  if (isSidePanelOpenSync() === false) {
    return false;
  }

  const response = await chrome.runtime.sendMessage<
    { type: string },
    boolean | undefined
  >({ type: PING_MESSAGE });
  return Boolean(response); // TODO: Drop Boolean() after strictNullChecks migration
}

export function getPopoverUrl(tabUrl: string | undefined): string | null {
  const popoverUrl = browser.runtime.getURL("restrictedUrlPopup.html");

  if (tabUrl?.startsWith(getExtensionConsoleUrl())) {
    return `${popoverUrl}?reason=${DISPLAY_REASON_EXTENSION_CONSOLE}`;
  }

  if (!isScriptableUrl(tabUrl)) {
    return `${popoverUrl}?reason=${DISPLAY_REASON_RESTRICTED_URL}`;
  }

  // The popup is disabled, and the extension will receive browserAction.onClicked events.
  // https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/browserAction/setPopup#popup
  return null;
}

export function getSidebarPath(tabId: number, url: string): string {
  return getPopoverUrl(url) ?? "sidebar.html?tabId=" + tabId;
}

export async function openSidePanel(tabId: number, url: string) {
  // Simultaneously define, enable, and open the side panel
  void chrome.sidePanel.setOptions({
    tabId,
    path: getSidebarPath(tabId, url),
    enabled: true,
  });
  await chrome.sidePanel.open({ tabId });
}

export async function hideSidePanel(tabId: number) {
  void chrome.sidePanel.setOptions({
    tabId,
    enabled: false,
  });
}

// Approximate sidebar width in pixels. Used to determine whether it's open
const MINIMUM_SIDEBAR_WIDTH = 300;

/**
 * Determines whether the sidebar is open.
 * @returns false when it's definitely closed
 * @returns 'unknown' when it cannot be determined
 */
// The type cannot be `undefined` due to strictNullChecks
function isSidePanelOpenSync(): false | "unknown" {
  if (!globalThis.window) {
    return "unknown";
  }

  return window.outerWidth - window.innerWidth > MINIMUM_SIDEBAR_WIDTH
    ? "unknown"
    : false;
}

// TODO: It doesn't work when the dev tools are open on the side
// Official event requested in https://github.com/w3c/webextensions/issues/517
export function onSidePanelClosure(controller: AbortController): void {
  expectContext("contentScript");
  window.addEventListener(
    "resize",
    () => {
      if (isSidePanelOpenSync() === false) {
        controller.abort();
      }
    },
    { signal: controller.signal },
  );
}
