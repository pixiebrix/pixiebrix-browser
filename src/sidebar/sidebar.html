<!--
  ~ Copyright (C) 2022 PixieBrix, Inc.
  ~
  ~ This program is free software: you can redistribute it and/or modify
  ~ it under the terms of the GNU Affero General Public License as published by
  ~ the Free Software Foundation, either version 3 of the License, or
  ~ (at your option) any later version.
  ~
  ~ This program is distributed in the hope that it will be useful,
  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  ~ GNU Affero General Public License for more details.
  ~
  ~ You should have received a copy of the GNU Affero General Public License
  ~ along with this program.  If not, see <http://www.gnu.org/licenses/>.
  -->

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <!-- Open all links in new tab, never in the sidebar #851 -->
    <base target="_blank" />
    <title>PixieBrix: Page Panel</title>
    <link rel="stylesheet" href="loadingScreen.css" />
    <link rel="stylesheet" href="sidebar.css" />
  </head>
  <body>
    <div id="container" class="full-height vh-100">
      <div class="global-loading-message">Loading Sidebar ðŸš€</div>
    </div>
    <script src="loadingScreen.js"></script>
    <script src="sidebar.js"></script>
    <!-- Must be at the bottom https://github.com/pixiebrix/pixiebrix-extension/issues/4100 -->
    <link
      href="https://fonts.googleapis.com/css2?family=Ubuntu:wght@300;400;500;700&display=swap"
      rel="stylesheet"
    />
  </body>
</html>
