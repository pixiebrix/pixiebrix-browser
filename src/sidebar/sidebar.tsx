/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// Keep in order so precedence is preserved
import "@/vendors/theme/app/app.scss";
import "@/vendors/overrides.scss";
import "@/utils/layout.scss";
import "./sidebar.scss";

import "@/extensionContext";

import registerMessenger from "@/sidebar/messenger/registration";
import App from "@/sidebar/SidebarApp";
import ReactDOM from "react-dom";
import React from "react";
import registerBuiltinBlocks from "@/blocks/registerBuiltinBlocks";
import registerContribBlocks from "@/contrib/registerContribBlocks";
import { initToaster } from "@/utils/notify";

registerMessenger();
registerContribBlocks();
registerBuiltinBlocks();
initToaster();

ReactDOM.render(<App />, document.querySelector("#container"));
