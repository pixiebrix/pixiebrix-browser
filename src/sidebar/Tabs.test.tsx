/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import Tabs from "@/sidebar/Tabs";
import { render } from "@/sidebar/testHelpers";
import { type SidebarEntries } from "@/types/sidebarTypes";
import sidebarSlice from "@/sidebar/sidebarSlice";
import { sidebarEntryFactory } from "@/testUtils/factories/sidebarEntryFactories";
import { screen, within } from "@testing-library/react";
import { MOD_LAUNCHER } from "@/sidebar/modLauncher/constants";
import { waitForEffect } from "@/testUtils/testHelpers";
import userEvent from "@testing-library/user-event";
import * as messengerApi from "@/contentScript/messenger/api";
import { eventKeyForEntry } from "@/sidebar/eventKeyUtils";

const cancelFormSpy = jest.spyOn(messengerApi, "cancelForm");
const hideSidebarSpy = jest.spyOn(messengerApi, "hideSidebar");

async function setupPanelsAndRender(options: {
  sidebarEntries?: Partial<SidebarEntries>;
  showModLauncher?: boolean;
  reservedSidebarEntries?: Partial<SidebarEntries>;
}) {
  const {
    sidebarEntries = {},
    showModLauncher = true,
    reservedSidebarEntries = {},
  } = options;
  (messengerApi.getReservedSidebarEntries as jest.Mock).mockImplementation(
    () => ({
      panels: [],
      temporaryPanels: [],
      forms: [],
      modActivationPanel: null,
      ...reservedSidebarEntries,
    })
  );

  const renderResult = render(<Tabs />, {
    setupRedux(dispatch) {
      dispatch(
        sidebarSlice.actions.setInitialPanels({
          panels: [],
          staticPanels: [],
          temporaryPanels: [],
          forms: [],
          modActivationPanel: undefined,
          ...sidebarEntries,
        })
      );

      if (!showModLauncher) {
        dispatch(sidebarSlice.actions.closeTab(eventKeyForEntry(MOD_LAUNCHER)));
      }
    },
  });

  await waitForEffect();

  return renderResult;
}

jest.mock("@/contentScript/messenger/api", () => ({
  ...jest.requireActual("@/contentScript/messenger/api"),
  getReservedSidebarEntries: jest.fn(),
}));

describe("Tabs", () => {
  const panel = sidebarEntryFactory("panel");

  test("renders", () => {
    const { asFragment } = render(<Tabs />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders with panels", async () => {
    await setupPanelsAndRender({ sidebarEntries: { panels: [panel] } });

    expect(screen.getByText("Panel Test 1")).toBeInTheDocument();
  });

  describe("Mod Launcher", () => {
    test("renders with mod launcher", async () => {
      const { asFragment } = await setupPanelsAndRender({
        sidebarEntries: {
          staticPanels: [MOD_LAUNCHER],
        },
      });

      expect(asFragment()).toMatchSnapshot();
    });

    test("displays no active sidebar panels view when no panels are active", async () => {
      await setupPanelsAndRender({
        sidebarEntries: { staticPanels: [MOD_LAUNCHER] },
      });

      expect(
        screen.getByText("We didn't find any mods to run")
      ).toBeInTheDocument();
      expect(
        screen.getByText("Don't worry, there's a solution")
      ).toBeInTheDocument();
    });

    test("can close the mod launcher", async () => {
      await setupPanelsAndRender({
        sidebarEntries: {
          panels: [panel],
          staticPanels: [MOD_LAUNCHER],
        },
      });

      expect(screen.getByText("Mods")).toBeInTheDocument();

      within(screen.getByRole("tab", { name: /mods/i }))
        .getByRole("button", { name: "Close" })
        .click();

      expect(
        screen.queryByRole("tab", { name: /mods/i })
      ).not.toBeInTheDocument();
    });

    test("can open the mod launcher", async () => {
      await setupPanelsAndRender({
        sidebarEntries: {
          panels: [panel],
          staticPanels: [MOD_LAUNCHER],
        },
      });

      expect(screen.getByText("Mods")).toBeInTheDocument();

      within(screen.getByRole("tab", { name: /mods/i }))
        .getByRole("button", { name: "Close" })
        .click();

      expect(screen.queryByText("Mods")).not.toBeInTheDocument();

      screen.getByRole("button", { name: "open mod launcher" }).click();

      expect(screen.getByText("Mods")).toBeInTheDocument();
    });

    test("clicking open the mod launcher multiple times does not open multiple mod launchers", async () => {
      await setupPanelsAndRender({
        sidebarEntries: { staticPanels: [MOD_LAUNCHER] },
      });

      expect(screen.getByText("Mods")).toBeInTheDocument();

      await userEvent.click(
        screen.getByRole("button", { name: "open mod launcher" })
      );

      expect(screen.getAllByText("Mods")).toHaveLength(1);
    });
  });

  describe("Persistent Panels", () => {
    (messengerApi.getReservedSidebarEntries as jest.Mock).mockImplementation(
      () => ({
        panels: [],
        temporaryPanels: [],
        forms: [],
        modActivationPanel: null,
      })
    );

    test("can close a panel when mod launcher is available", async () => {
      await setupPanelsAndRender({
        sidebarEntries: {
          panels: [panel],
          staticPanels: [MOD_LAUNCHER],
        },
      });

      within(screen.getByRole("tab", { name: /panel test 1/i }))
        .getByRole("button", { name: "Close" })
        .click();

      expect(
        screen.queryByRole("tab", { name: /panel test 1/i })
      ).not.toBeInTheDocument();
    });

    test("can open a closed panel when mod launcher is available", async () => {
      await setupPanelsAndRender({
        sidebarEntries: {
          panels: [panel],
          staticPanels: [MOD_LAUNCHER],
        },
      });

      within(screen.getByRole("tab", { name: /panel test 1/i }))
        .getByRole("button", { name: "Close" })
        .click();

      expect(
        screen.queryByRole("tab", { name: /panel test 1/i })
      ).not.toBeInTheDocument();

      screen.getByRole("heading", { name: /panel test 1/i }).click();

      expect(
        screen.getByRole("tab", { name: /panel test 1/i })
      ).toBeInTheDocument();
    });

    test("sidebar doesn't close if visible reserved panel available", async () => {
      hideSidebarSpy.mockReset();
      await setupPanelsAndRender({
        sidebarEntries: {
          panels: [],
        },
        reservedSidebarEntries: { panels: [panel] },
      });

      await waitForEffect();
      expect(hideSidebarSpy).toHaveBeenCalledTimes(0);
    });

    test("sidebar closes if no visible reserved panel available", async () => {
      hideSidebarSpy.mockReset();
      await setupPanelsAndRender({
        sidebarEntries: {
          panels: [],
        },
      });

      await waitForEffect();
      expect(hideSidebarSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe("Temporary Panels", () => {
    const temporaryPanel = sidebarEntryFactory("temporaryPanel");

    test("can close a temporary panel when mod launcher is available", async () => {
      await setupPanelsAndRender({
        sidebarEntries: {
          temporaryPanels: [temporaryPanel],
          staticPanels: [MOD_LAUNCHER],
        },
      });

      within(screen.getByRole("tab", { name: /temporary panel test 1/i }))
        .getByRole("button", { name: "Close" })
        .click();

      expect(
        screen.queryByRole("tab", { name: /temporary panel test 1/i })
      ).not.toBeInTheDocument();
    });

    test("can close a tempoarary panel when mod launcher is unavailable", async () => {
      await setupPanelsAndRender({
        sidebarEntries: { temporaryPanels: [temporaryPanel] },
      });

      within(screen.getByRole("tab", { name: /temporary panel test 1/i }))
        .getByRole("button", { name: "Close" })
        .click();

      expect(
        screen.queryByRole("tab", { name: /temporary panel test 1/i })
      ).not.toBeInTheDocument();
    });

    test("cannot re-open a closed temporary panel with the mod launcher", async () => {
      await setupPanelsAndRender({
        sidebarEntries: {
          temporaryPanels: [temporaryPanel],
          staticPanels: [MOD_LAUNCHER],
        },
      });

      within(screen.getByRole("tab", { name: /temporary panel test 1/i }))
        .getByRole("button", { name: "Close" })
        .click();

      expect(
        screen.queryByRole("tab", { name: /temporary panel test 1/i })
      ).not.toBeInTheDocument();

      expect(
        screen.queryByRole("heading", { name: /temporary panel test 1/i })
      ).not.toBeInTheDocument();
    });

    test("sidebar doesn't close if visible reserved panel available", async () => {
      hideSidebarSpy.mockReset();
      await setupPanelsAndRender({
        sidebarEntries: {
          temporaryPanels: [],
        },
        reservedSidebarEntries: { temporaryPanels: [temporaryPanel] },
      });

      await waitForEffect();
      expect(hideSidebarSpy).toHaveBeenCalledTimes(0);
    });

    test("sidebar closes if no visible reserved panel available", async () => {
      hideSidebarSpy.mockReset();
      await setupPanelsAndRender({
        sidebarEntries: {
          temporaryPanels: [],
        },
      });

      await waitForEffect();
      expect(hideSidebarSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe("Form Panels", () => {
    const formPanel = sidebarEntryFactory("form");

    test("renders with forms", async () => {
      await setupPanelsAndRender({ sidebarEntries: { forms: [formPanel] } });

      expect(screen.getByText("Form Panel Test")).toBeInTheDocument();
    });

    test("closing the form tab calls cancelForm", async () => {
      await setupPanelsAndRender({
        sidebarEntries: {
          forms: [formPanel],
          staticPanels: [MOD_LAUNCHER],
        },
      });

      expect(cancelFormSpy).not.toHaveBeenCalled();

      within(screen.getByRole("tab", { name: /form panel test/i }))
        .getByRole("button", { name: "Close" })
        .click();

      await waitForEffect();

      expect(cancelFormSpy).toHaveBeenCalledWith(
        {
          frameId: 0,
          tabId: 1,
        },
        formPanel.nonce
      );
    });
  });

  describe("Activation Panels", () => {
    const activatePanel = sidebarEntryFactory("activateMods");

    test("close the activation panel and make mod launcher active", async () => {
      await setupPanelsAndRender({
        sidebarEntries: {
          staticPanels: [MOD_LAUNCHER],
          modActivationPanel: activatePanel,
        },
      });

      within(screen.getByRole("tab", { name: /activate mods test 1/i }))
        .getByRole("button", { name: "Close" })
        .click();

      expect(
        screen.queryByRole("tab", { name: /activate mods test 1/i })
      ).not.toBeInTheDocument();

      expect(screen.queryByRole("tab", { name: /mods/i })).toHaveClass(
        "active"
      );
    });

    test("closing the activation panel hides sidebar if it's the only open panel", async () => {
      hideSidebarSpy.mockReset();
      await setupPanelsAndRender({
        sidebarEntries: {
          staticPanels: [MOD_LAUNCHER],
          modActivationPanel: activatePanel,
        },
        showModLauncher: false,
      });

      within(screen.getByRole("tab", { name: /activate mods test 1/i }))
        .getByRole("button", { name: "Close" })
        .click();

      expect(
        screen.queryByRole("tab", { name: /activate mods test 1/i })
      ).not.toBeInTheDocument();

      await waitForEffect();

      expect(hideSidebarSpy).toHaveBeenCalledTimes(1);
    });
  });
});
