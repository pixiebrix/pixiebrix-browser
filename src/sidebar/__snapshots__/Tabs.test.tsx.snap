// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Mod Launcher renders with mod launcher 1`] = `
<DocumentFragment>
  <div
    class="full-height"
  >
    <div
      class="tabContainer nav nav-tabs nav-justified"
      role="tablist"
    >
      <div
        class="tabWrapper active nav-item"
      >
        <a
          aria-controls="panel-container-tabpane-static-modLauncher-panel"
          aria-selected="true"
          class="tabHeader nav-link active"
          data-rb-event-key="static-modLauncher-panel"
          href="#"
          id="panel-container-tab-static-modLauncher-panel"
          role="tab"
          target="_self"
        >
          <span
            class="tabTitle"
          >
            Mods
          </span>
          <button
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
            <span
              class="sr-only"
            >
              Close
            </span>
          </button>
        </a>
        <div
          class="tabDivider"
        />
      </div>
      <button
        aria-label="open mod launcher"
        class="addButton btn btn-link btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-plus fa-w-14 "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="p-0 border-0 full-height bg-white tab-content"
    >
      <div
        aria-hidden="false"
        aria-labelledby="panel-container-tab-static-modLauncher-panel"
        class="fade h-100 paneOverrides tab-pane active show"
        id="panel-container-tabpane-static-modLauncher-panel"
        role="tabpanel"
      >
        <div
          class="full-height h-100"
        >
          <div
            class="scrollable-area container"
          >
            <div
              class="emptyViewRoot"
            >
              <div>
                <p>
                  We didn't find any mods to run
                </p>
                <h4>
                  Don't worry, there's a solution
                </h4>
                <img
                  alt="Workshop"
                  class="illustration"
                  src="test-file-stub"
                />
                <p>
                  Some mods don't run on every page.
                  <br />
                  And other mods don't run in the Sidebar.
                  <br />
                  <br />
                  Check that the mod is configured correctly.
                </p>
              </div>
              <p>
                Looking for new mods?
                <br />
                Check out the 
                <a
                  href="https://www.pixiebrix.com/marketplace/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  PixieBrix Marketplace
                </a>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs renders 1`] = `
<DocumentFragment>
  <div
    class="full-height"
  >
    <div
      class="tabContainer nav nav-tabs nav-justified"
      role="tablist"
    >
      <button
        aria-label="open mod launcher"
        class="addButton btn btn-link btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-plus fa-w-14 "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="p-0 border-0 full-height bg-white tab-content"
    />
  </div>
</DocumentFragment>
`;
